ffi_internals/src/type_ffi.rs

```diff
@@ -0,0 +1,530 @@
+                    } else {
+                        (quote!((*Box::from_raw(#field_name)).into()), quote!(None))
+                    };
+                    if self.is_option {
+                        quote! {
+                            if #field_name.is_null() {
+                                #none
+                            } else {
+                                #conversion_or_borrow
+                            }
+                        }
+                    } else {
+                        quote! {
+                            #conversion_or_borrow
+                        }
+                    }
+                } else if self.is_option {
+                    let (conversion_or_borrow, none) = if self.is_borrow {
+                        (quote!(Some(&*#field_name)), quote!(&None))
+                    } else {
+                        (quote!(Some(*Box::from_raw(#field_name))), quote!(None))
+                    };
+                    quote! {
+                        if #field_name.is_null() {
+                            #none
+                        } else {
+                            #conversion_or_borrow
+                        }
+                    }
+                } else {
+                    let conversion_or_borrow = if self.is_borrow {
+                        quote!(&*#field_name)
+                    } else {
+                        quote!(*Box::from_raw(#field_name))
+                    };
+                    quote!(#conversion_or_borrow)
+                }
+            }
+            TypeIdentifier::DateTime => {
+                if self.is_option {
+                    quote! {
+                        if #field_name.is_null() {
+                            None
+                        } else {
+                            Some((&*Box::from_raw(#field_name)).into())
+                        }
+                    }
+                } else {
+                    quote!((&*Box::from_raw(#field_name)).into())
+                }
+            }
+            TypeIdentifier::Raw(_) => {
+                if self.is_option {
+                    quote! {
+                        if #field_name.is_null() {
+                            None
+                        } else {
+                            Some(*Box::from_raw(#field_name))
+                        }
+                    }
+                } else {
+                    quote!(#field_name)
+                }
+            }
+            TypeIdentifier::String => {
+                if self.is_option {
+                    quote! {
+                        if #field_name.is_null() {
+                            None
+                        } else {
+                            Some(ffi_common::core::string::string_from_c(#field_name))
+                        }
+                    }
+                } else {
+                    quote!(ffi_common::core::string::string_from_c(#field_name))
+                }
+            }
```

Comment:
Do these sections read nicer if you include the `if` with the match? (`TypeIdentifier::String if self.is_option =>`)

Reply:
Oh yeah, flattening that makes it a bit easier to follow. It's still not pretty, but each section is cleaner 👍